File: packages/widget/test/header-menu.test.ts
// @vitest-environment jsdom
//
// The header ⋯ menu, mounted the way the widget really mounts it: inside a
// shadow root. That detail is the entire point of this file. The menu's
// outside-dismiss listener lives on the DOCUMENT, and to a document-level
// listener every event from inside a shadow tree is retargeted to the shadow
// HOST — so a press on one of the menu's own items used to look "outside",
// close the menu mid-press, and swallow the click those pointer events were
// producing. Every item then "did nothing" under a real pointer while a
// synthetic `.click()` (no pointerdown) sailed through, which is how the menu
// passed automation and failed a human (dh-hyperlocal, reported issue 5).
// A test that appends the menu straight to `document.body` cannot catch any
// of this, because without a shadow boundary nothing is retargeted.

import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { createHeaderMenu, type HeaderMenuCallbacks } from '../src/ui/header-menu.js';

let host: HTMLElement;
let shadow: ShadowRoot;

beforeEach(() => {
  host = document.createElement('div');
  shadow = host.attachShadow({ mode: 'open' });
  document.body.appendChild(host);
});

afterEach(() => {
  host.remove();
  vi.restoreAllMocks();
});

function build(overrides: Partial<HeaderMenuCallbacks> = {}) {
  const callbacks: HeaderMenuCallbacks = {
    onStartNew: vi.fn(),
    onEndConversation: vi.fn(),
    onReportIssue: vi.fn(),
    onMuteChange: vi.fn(),
    ...overrides,
  };
  const menu = createHeaderMenu(callbacks);
  shadow.appendChild(menu.node);
  menu.update({ canEnd: true, privacyUrl: '', reportIssue: true, muted: false });
  return { menu, callbacks };
}

/** A real pointer's press: bubbles and crosses the shadow boundary. */
const press = (target: EventTarget) =>
  target.dispatchEvent(new Event('pointerdown', { bubbles: true, composed: true }));

const toggle = () => shadow.querySelector<HTMLButtonElement>('.dh-hmenu-toggle')!;
const menuEl = () => shadow.querySelector<HTMLElement>('.dh-hmenu')!;
/** The mute item — the first row in the menu, and the only one that toggles. */
const muteItem = () => menuEl().querySelector<HTMLButtonElement>('.dh-hmenu-item')!;

describe('outside-dismiss vs shadow retargeting', () => {
  it('sanity: the document really does see our presses retargeted to the host', () => {
    // If jsdom ever stops retargeting, the two tests below go vacuous — this
    // one fails first and says why.
    build();
    toggle().click();
    let seen: EventTarget | null = null;
    const spy = (event: Event): void => {
      seen = event.target;
    };
    document.addEventListener('pointerdown', spy);
    press(muteItem());
    document.removeEventListener('pointerdown', spy);
    expect(seen).toBe(host);
  });

  it('a press on a menu item does NOT close the menu, so its click still lands', () => {
    const { callbacks } = build();
    toggle().click();
    expect(menuEl().hidden).toBe(false);

    const mute = muteItem();
    press(mute);
    // Still open at "release" time — the click this press produces can land.
    expect(menuEl().hidden).toBe(false);

    mute.click();
    expect(callbacks.onMuteChange).toHaveBeenCalledWith(true);
    expect(menuEl().hidden).toBe(true);
  });

  it('a press elsewhere in the shadow tree (not just the host page) closes it', () => {
    build();
    const elsewhere = document.createElement('button');
    shadow.appendChild(elsewhere);
    toggle().click();

    press(elsewhere);
    expect(menuEl().hidden).toBe(true);
  });

  it('a press on the host page outside the widget closes it', () => {
    build();
    toggle().click();

    press(document.body);
    expect(menuEl().hidden).toBe(true);
  });
});

// ── The mute item's label (reported issue 2) ──────────────────────────────
//
// It used to read "Mute notifications" in both states — only the bell glyph
// changed — so a customer who had muted the chime was offered "Mute
// notifications" again and could not tell what pressing it would do. The item
// is now a plain `menuitem` whose label names the ACTION, which is why the
// `menuitemcheckbox` role and `aria-checked` had to go with it: a checkbox
// announcing "Unmute notifications, checked" states the opposite of what the
// control does. See `setMuted`'s own comment in ui/header-menu.ts.
describe('the mute item states what pressing it will do', () => {
  it('reads "Mute notifications" while sound is on and "Unmute notifications" once muted', () => {
    const { menu } = build();
    toggle().click();
    expect(muteItem().textContent).toBe('Mute notifications');

    muteItem().click();
    toggle().click();
    expect(muteItem().textContent).toBe('Unmute notifications');

    // And back — the label is derived from the state, not toggled blindly.
    muteItem().click();
    toggle().click();
    expect(muteItem().textContent).toBe('Mute notifications');
    menu.close();
  });

  it('renders the muted label from an `update()` that arrives already muted', () => {
    // The persisted preference path: the widget reads the stored mute flag and
    // pushes it in through `update`, which must paint the same label a click
    // would have.
    const { menu } = build();
    menu.update({ canEnd: true, privacyUrl: '', reportIssue: true, muted: true });
    toggle().click();
    expect(muteItem().textContent).toBe('Unmute notifications');
  });

  it('is a plain menuitem with no contradictory checked state', () => {
    build();
    toggle().click();
    expect(muteItem().getAttribute('role')).toBe('menuitem');
    expect(muteItem().hasAttribute('aria-checked')).toBe(false);
    // The accessible name IS the visible label — nothing overrides it, so
    // WCAG 2.5.3 (Label in Name) holds in both states.
    expect(muteItem().hasAttribute('aria-label')).toBe(false);

    muteItem().click();
    toggle().click();
    expect(muteItem().hasAttribute('aria-checked')).toBe(false);
    expect(muteItem().hasAttribute('aria-label')).toBe(false);
  });
});

File: packages/flutter/test/ui/header/header_menu_test.dart
import 'package:dhaam_chat_flutter/dhaam_chat_flutter.dart';
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';
import 'package:flutter_test/flutter_test.dart';

/// Reproduces `packages/widget/test/header-menu.test.ts` and the
/// `the conversation menu` block of `remote-config-gating.test.ts`.
///
/// Four of `header-menu.test.ts`'s seven cases are its `outside-dismiss vs
/// shadow retargeting` block, which does not transfer — see `header_menu.dart`
/// for why (no shadow boundary, so nothing to retarget). The dismissal
/// behaviour those cases were protecting is asserted below through the
/// framework's own modal barrier instead of around it.
void main() {
  Future<void> mount(
    WidgetTester tester, {
    bool canEnd = true,
    String? privacyUrl,
    bool reportIssue = true,
    bool muted = false,
    ValueChanged<bool>? onMuteChange,
    VoidCallback? onStartNew,
    VoidCallback? onEndConversation,
    VoidCallback? onReportIssue,
    ValueChanged<String>? onOpenPrivacy,
  }) {
    return tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            actions: <Widget>[
              HeaderMenu(
                canEnd: canEnd,
                privacyUrl: privacyUrl,
                reportIssue: reportIssue,
                muted: muted,
                onStartNew: onStartNew ?? () {},
                onEndConversation: onEndConversation ?? () {},
                onReportIssue: onReportIssue ?? () {},
                onMuteChange: onMuteChange ?? (_) {},
                onOpenPrivacy: onOpenPrivacy ?? (_) {},
              ),
            ],
          ),
        ),
      ),
    );
  }

  Future<void> openMenu(WidgetTester tester) async {
    await tester.tap(find.byIcon(Icons.more_horiz));
    await tester.pumpAndSettle();
  }

  /// Every row label currently on screen, in order — the Dart counterpart of
  /// the TS helper that maps `.dh-hmenu-item` textContent, filtering hidden.
  /// Nothing is filtered here, because an unbacked row is never built.
  List<String> labels(WidgetTester tester) => tester
      .widgetList<PopupMenuItem<HeaderMenuAction>>(
        find.byType(PopupMenuItem<HeaderMenuAction>),
      )
      .map(
        (PopupMenuItem<HeaderMenuAction> item) =>
            (((item.child! as Row).children.last as Expanded).child
                    as Text)
                .data!,
      )
      .toList();

  // ── The mute item states what pressing it will do ─────────────────────
  //
  // It used to read "Mute notifications" in both states — only the bell glyph
  // changed — so a customer who had muted the chime was offered "Mute
  // notifications" again and could not tell what pressing it would do.
  group('the mute item states what pressing it will do', () {
    test('the label is derived from the state, never toggled blindly', () {
      // The pure half, with no widget at all: two states, two strings, and a
      // function that cannot drift out of step with itself.
      expect(muteLabel(muted: false), 'Mute notifications');
      expect(muteLabel(muted: true), 'Unmute notifications');
    });

    testWidgets(
      'reads "Mute notifications" while sound is on and "Unmute '
      'notifications" once muted',
      (WidgetTester tester) async {
        bool muted = false;
        await tester.pumpWidget(
          MaterialApp(
            home: StatefulBuilder(
              builder: (BuildContext context, StateSetter setState) => Scaffold(
                appBar: AppBar(
                  actions: <Widget>[
                    HeaderMenu(
                      canEnd: true,
                      privacyUrl: null,
                      reportIssue: true,
                      muted: muted,
                      onStartNew: () {},
                      onEndConversation: () {},
                      onReportIssue: () {},
                      onMuteChange: (bool next) =>
                          setState(() => muted = next),
                      onOpenPrivacy: (_) {},
                    ),
                  ],
                ),
              ),
            ),
          ),
        );

        await openMenu(tester);
        expect(find.text('Mute notifications'), findsOneWidget);

        await tester.tap(find.text('Mute notifications'));
        await tester.pumpAndSettle();

        await openMenu(tester);
        expect(find.text('Unmute notifications'), findsOneWidget);
        expect(find.text('Mute notifications'), findsNothing);

        // And back — the label follows the state in both directions.
        await tester.tap(find.text('Unmute notifications'));
        await tester.pumpAndSettle();
        await openMenu(tester);
        expect(find.text('Mute notifications'), findsOneWidget);
      },
    );

    testWidgets('paints the muted label from a build that arrives already '
        'muted', (WidgetTester tester) async {
      // The persisted-preference path: a stored mute flag pushed straight in,
      // which must paint the same label a tap would have. Here it is not a
      // special path at all — `muted` is a prop, so there is no second copy
      // that could be out of date.
      await mount(tester, muted: true);
      await openMenu(tester);
      expect(find.text('Unmute notifications'), findsOneWidget);
    });

    testWidgets('reports the NEW muted state, not the old one',
        (WidgetTester tester) async {
      final List<bool> reported = <bool>[];
      await mount(tester, muted: false, onMuteChange: reported.add);
      await openMenu(tester);
      await tester.tap(find.text('Mute notifications'));
      await tester.pumpAndSettle();
      expect(reported, <bool>[true]);
    });

    testWidgets('reports false when a muted visitor asks for sound back',
        (WidgetTester tester) async {
      final List<bool> reported = <bool>[];
      await mount(tester, muted: true, onMuteChange: reported.add);
      await openMenu(tester);
      await tester.tap(find.text('Unmute notifications'));
      await tester.pumpAndSettle();
      expect(reported, <bool>[false]);
    });

    testWidgets(
      'is a plain menu item with no contradictory checked state',
      (WidgetTester tester) async {
        // The a11y contract that replaced `menuitemcheckbox`/`aria-checked`.
        // A checkbox announcing "Unmute notifications, checked" would state
        // the opposite of what the control does; pinning a stable accessible
        // name instead is a WCAG 2.5.3 (Label in Name) failure.
        // Disposed inside the body rather than through `addTearDown`: the
        // framework's end-of-test verification runs BEFORE tear-downs and
        // fails on a live handle.
        final SemanticsHandle handle = tester.ensureSemantics();

        await mount(tester, muted: false);
        await openMenu(tester);

        SemanticsNode row(String label) =>
            tester.getSemantics(find.text(label));

        final SemanticsData unmutedRow = row('Mute notifications').getSemanticsData();
        expect(unmutedRow.hasFlag(SemanticsFlag.hasCheckedState), isFalse);
        expect(unmutedRow.hasFlag(SemanticsFlag.isChecked), isFalse);
        expect(unmutedRow.hasFlag(SemanticsFlag.isButton), isTrue);
        // The accessible name IS the visible label — nothing overrides it.
        expect(unmutedRow.label, 'Mute notifications');

        // Selecting a row pops the menu route, so the tree is clean before
        // the muted build is mounted over it.
        await tester.tap(find.text('Mute notifications'));
        await tester.pumpAndSettle();

        await mount(tester, muted: true);
        await openMenu(tester);

        final SemanticsData mutedRow =
            row('Unmute notifications').getSemanticsData();
        expect(mutedRow.hasFlag(SemanticsFlag.hasCheckedState), isFalse);
        expect(mutedRow.hasFlag(SemanticsFlag.isChecked), isFalse);
        expect(mutedRow.label, 'Unmute notifications');

        handle.dispose();
      },
    );
  });

  // ── Nothing in this menu is decorative ────────────────────────────────
  group('offers only what is actually backed', () {
    test('the entry list is decided by a pure function, off any widget tree',
        () {
      expect(
        headerMenuEntries(
          canEnd: false,
          privacyUrl: null,
          reportIssue: false,
          muted: false,
        ).map((HeaderMenuEntry e) => e.label),
        <String>['Mute notifications', 'Start new conversation'],
      );
    });

    testWidgets('drops End conversation when there is nothing live to end',
        (WidgetTester tester) async {
      await mount(tester, canEnd: false, reportIssue: false);
      await openMenu(tester);
      expect(labels(tester),
          <String>['Mute notifications', 'Start new conversation']);
      expect(find.text('End conversation'), findsNothing);
    });

    testWidgets('offers End conversation on a live conversation',
        (WidgetTester tester) async {
      await mount(tester, canEnd: true, reportIssue: false);
      await openMenu(tester);
      expect(labels(tester), contains('End conversation'));
    });

    testWidgets('offers nothing until the merchant turns Report on',
        (WidgetTester tester) async {
      await mount(tester, reportIssue: false);
      await openMenu(tester);
      expect(labels(tester), isNot(contains('Report an issue')));
    });

    testWidgets('adds Report an issue when the merchant offers it',
        (WidgetTester tester) async {
      await mount(tester, reportIssue: true);
      await openMenu(tester);
      expect(labels(tester), contains('Report an issue'));
    });

    testWidgets('hidden, never disabled — an unbacked row is not built at all',
        (WidgetTester tester) async {
      // The distinction the reference is explicit about: a disabled row still
      // reads as a feature the customer is not allowed to use, and there is
      // no permission to go looking for.
      await mount(tester, canEnd: false, reportIssue: false);
      await openMenu(tester);
      for (final PopupMenuItem<HeaderMenuAction> item
          in tester.widgetList<PopupMenuItem<HeaderMenuAction>>(
        find.byType(PopupMenuItem<HeaderMenuAction>),
      )) {
        expect(item.enabled, isTrue);
      }
    });
  });

  group('Privacy links the merchant policy, or is absent', () {
    testWidgets('hides Privacy until the merchant sets a URL',
        (WidgetTester tester) async {
      await mount(tester, privacyUrl: null);
      await openMenu(tester);
      expect(labels(tester), isNot(contains('Privacy')));
    });

    testWidgets('hides Privacy for a blank URL', (WidgetTester tester) async {
      await mount(tester, privacyUrl: '   ');
      await openMenu(tester);
      expect(labels(tester), isNot(contains('Privacy')));
    });

    testWidgets("links Privacy to the merchant's own policy",
        (WidgetTester tester) async {
      final List<String> opened = <String>[];
      await mount(
        tester,
        privacyUrl: 'https://acme.test/privacy',
        onOpenPrivacy: opened.add,
      );
      await openMenu(tester);
      expect(labels(tester), contains('Privacy'));

      await tester.tap(find.text('Privacy'));
      await tester.pumpAndSettle();
      expect(opened, <String>['https://acme.test/privacy']);
    });

    testWidgets('refuses a javascript: privacy URL rather than linking it',
        (WidgetTester tester) async {
      // Merchant-supplied and about to become a navigation, so the allowlist
      // has to apply — and it applies HERE, at the link, not at the parse.
      await mount(tester, privacyUrl: 'javascript:alert(1)');
      await openMenu(tester);
      expect(labels(tester), isNot(contains('Privacy')));
    });

    test('the offered href is exactly what safeLinkUrl accepted', () {
      final HeaderMenuEntry privacy = headerMenuEntries(
        canEnd: false,
        privacyUrl: '  https://acme.test/privacy  ',
        reportIssue: false,
        muted: false,
      ).last;
      expect(privacy.action, HeaderMenuAction.privacy);
      // Trimmed by the validator, not by this module — one answer, one place.
      expect(privacy.href, 'https://acme.test/privacy');
    });
  });

  group('the rest of the rows reach their callbacks', () {
    testWidgets('Start new conversation', (WidgetTester tester) async {
      int calls = 0;
      await mount(tester, onStartNew: () => calls += 1);
      await openMenu(tester);
      await tester.tap(find.text('Start new conversation'));
      await tester.pumpAndSettle();
      expect(calls, 1);
    });

    testWidgets('End conversation', (WidgetTester tester) async {
      int calls = 0;
      await mount(tester, onEndConversation: () => calls += 1);
      await openMenu(tester);
      await tester.tap(find.text('End conversation'));
      await tester.pumpAndSettle();
      expect(calls, 1);
    });

    testWidgets('Report an issue', (WidgetTester tester) async {
      int calls = 0;
      await mount(tester, onReportIssue: () => calls += 1);
      await openMenu(tester);
      await tester.tap(find.text('Report an issue'));
      await tester.pumpAndSettle();
      expect(calls, 1);
    });
  });

  group('opening and dismissing', () {
    testWidgets('starts closed', (WidgetTester tester) async {
      await mount(tester);
      expect(find.byType(PopupMenuItem<HeaderMenuAction>), findsNothing);
    });

    testWidgets('the toggle carries the reference name, not the framework '
        'default', (WidgetTester tester) async {
      await mount(tester);
      expect(
        tester
            .widget<PopupMenuButton<HeaderMenuAction>>(
              find.byType(PopupMenuButton<HeaderMenuAction>),
            )
            .tooltip,
        'Conversation options',
      );
    });

    testWidgets('a press outside closes it, and fires nothing',
        (WidgetTester tester) async {
      // The behaviour `header-menu.test.ts`'s retargeting block protects,
      // asserted through the framework's own modal barrier — there is no
      // shadow boundary here for a press to be retargeted across.
      final List<bool> muteCalls = <bool>[];
      await mount(tester, onMuteChange: muteCalls.add);
      await openMenu(tester);
      expect(find.text('Mute notifications'), findsOneWidget);

      await tester.tapAt(const Offset(20, 400));
      await tester.pumpAndSettle();

      expect(find.byType(PopupMenuItem<HeaderMenuAction>), findsNothing);
      expect(muteCalls, isEmpty);
    });

    testWidgets('a press on a row still lands — it is not swallowed by the '
        'dismiss', (WidgetTester tester) async {
      final List<bool> muteCalls = <bool>[];
      await mount(tester, onMuteChange: muteCalls.add);
      await openMenu(tester);
      await tester.tap(find.text('Mute notifications'));
      await tester.pumpAndSettle();
      expect(muteCalls, <bool>[true]);
      expect(find.byType(PopupMenuItem<HeaderMenuAction>), findsNothing);
    });
  });
}
